programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -804,6 +804,17 @@ fn translate_account_infos<'a, T, F>(
 where
     F: Fn(&T) -> u64,
 {
+    let direct_mapping = invoke_context
+        .get_feature_set()
+        .is_active(&feature_set::bpf_account_data_direct_mapping::id());
+
+    // In the same vein as the other check_account_info_pointer() checks, we don't lock
+    // this pointer to a specific address but we don't want it to be inside accounts, or
+    // callees might be able to write to the pointed memory.
+    if direct_mapping && account_infos_addr >= ebpf::MM_INPUT_START {
```

Comment:
This should check the end of the slice, not the beginning.

Reply:
True. This is also done in some other places, I'll fix those up too.

I suppose this would be hard to exploit but might as well do it right.

Reply:
Actually I couldn't not find any others

Reply:
So I've re-writtten this. It looks a bit like an open-coded llvm gep :) Let me know if there is a better way of doing this.